### PR TITLE
Fix GUI test failures due to overriding of __eq__ and __hash__ methods

### PR DESCRIPTION
**Pull request by kristjaneerik**
I figured out why I was seeing the GUI errors on `master` when run together but not when run separately: some of the tests change global state by overriding the `__eq__` and `__hash__` methods of certain classes at the beginning of a test, but do not change them back at the end.

I've created a context manager to easily do both setting and resetting, and have applied to everywhere I saw it done.

The test suite now only has two sporadic test failures, and I'm 99% certain they are due to race conditions. This includes the `FAIL: testChangeScale (test_gaussian_1d.TestGaussian1D)` failure I thought I introduced in #269. The other one I see is `testPlotArgs (test_histogram.TestHistogram)`. Both fail when run on their own, but not all of the time. They've never failed when a delay like a `print()` is put right in front of the failing assertion.

### CODE REVIEW

**Comment by bpteague** [comment]
I'll have a look at this and your other PR in the next few days. I really appreciate your contributions! (:

**Comment by kristjaneerik** [comment]
Thanks, I hope they're useful! I'll work on merging in master to the various PRs.

In this branch I see one error, but I also see that on master:
```
======================================================================
ERROR: testPlotParams (test_density_plot.TestDensityPlot)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eerik/Documents/cytoflow-dev/cytoflowgui/tests/test_density_plot.py", line 251, in testPlotParams
    wait_for_condition(lambda v: v.view_error == "", self.wi, 'view_error', 30)
  File "/home/eerik/miniconda3/envs/cf_dev/lib/python3.6/site-packages/traits/util/async_trait_wait.py", line 62, in wait_for_condition
    raise RuntimeError("Timed out waiting for condition.")
RuntimeError: Timed out waiting for condition.

-------------------- >> begin captured logging << --------------------
cytoflowgui.workflow: ERROR: Error: TypeError: '<' not supported between instances of 'function' and 'function'

Location:   File "/home/eerik/miniconda3/envs/cf_dev/lib/python3.6/queue.py", line 230, in _get
    return heappop(self.queue)
Thread: MainThread
--------------------- >> end captured logging << ---------------------
```

I don't get the `ERROR: testNotebook (test_bleedthrough_linear.TestBleedthroughLinear)` here though that I get on master.

File: cytoflowgui/tests/test_tasbe.py [hunk 1]
```diff
@@ -167,9 +160,8 @@ def testNotebook(self):
 
         nb_data = locals()['ex_2'].data
         remote_data = self.workflow.remote_eval("self.workflow[-1].result.data")
-        
-        self.assertTrue((nb_data == remote_data).all().all())
+        pd.testing.assert_frame_equal(nb_data, remote_data)
```
Comment (kristjaneerik) [inline cytoflowgui/tests/test_tasbe.py:163]: Had to change to using this function for the testing -- it does a slightly fuzzy comparison: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.testing.assert_frame_equal.html

**Comment by bpteague** [comment]
That's a subtle bug -- nice catch! And the context manager is a very elegant solution.

I resisted adding unit tests for the GUI for a long time - they're an ugly PITA and subject to a bunch of subtle race conditions that I've mostly gotten under control. But on the other hand, they uncovered a TON of bugs in the GUI and made it a lot more stable.

Which is to say, thanks so much. Merged (: :+1:

**Pull request merged**